### DISCUSSION TITLE
General Performance tips

### DISCUSSION THREAD

**Discussion by mrsquish** [QuestPDF/QuestPDF#336] [Q&A] [answered]
My use case for QuestPDF will typically generate PDFs with hundreds of pages. I'm trying to shave the rendering times down as much as possible. Are there any general tips for improving performance? e.g. layout components that are faster than others, Fixed sizes vs variable sizes, particular properties that cause slowness etc.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Since this was posted, QuestPDF has undergone significant performance improvements. A few general tips for large documents: use image optimization via `WithRasterDpi()` and `WithCompressionQuality()` to reduce file size and rendering time, load images from file paths to take advantage of automatic shared image caching, and consider using the `Lazy` element for content-heavy documents to defer rendering of off-screen sections.

More details can be found in the documentation:
https://www.questpdf.com/api-reference/image/optimization.html
https://www.questpdf.com/api-reference/image/shared.html

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
